common/overlay/overlay.go

```diff
@@ -34,6 +36,33 @@ type MountCfg struct {
 	Lbl string
 }
 
+func DirSupportsOverlay(path string) (bool, error) {
+	dir, err := os.OpenFile(path, syscall.O_RDONLY|syscall.O_DIRECTORY, 0755)
+	if err != nil {
+		return false, errwrap.Wrap(fmt.Errorf("cannot open %q", dir), err)
+	}
+	defer dir.Close()
+
+	readDirentBuf := make([]byte, 4096)
+	var n int
+	n, err = syscall.ReadDirent(int(dir.Fd()), readDirentBuf)
+	if err != nil {
+		return false, errwrap.Wrap(fmt.Errorf("cannot read directory %q", dir), err)
+	}
+	readDirentOffset := 0
+	for readDirentOffset < n {
+		dirent := (*syscall.Dirent)(unsafe.Pointer(&readDirentBuf[readDirentOffset]))
```

Comment:
Good news though, `syscall.ReadDirent` forwards to the raw http://man7.org/linux/man-pages/man2/getdents.2.html syscall, passing the length of the buffer. The syscall maintains the invariant of not exceeding the buffer size.